compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -21,6 +22,98 @@ use super::{
 };
 use crate::ty;
 
+/// Functionality required for the bytes of an `Allocation`.
+pub trait AllocBytes:
+    Clone + core::fmt::Debug + Eq + PartialEq + PartialOrd + Ord + core::hash::Hash
+{
+    /// The length of the bytes.
+    fn get_len(&self) -> usize;
+    /// The address of the bytes.
+    fn expose_addr(&self) -> u64;
+    /// Get a slice of the bytes corresponding to a specified range.
+    fn get_slice_from_range(&self, range: Range<usize>) -> &[u8];
+    /// Mutable slice of the bytes corresponding to a specified range.
+    fn get_slice_from_range_mut<'a>(&'a mut self, range: Range<usize>) -> &'a mut [u8];
+    /// Add to the pointer of the head of the bytes, and return a mutable pointer to this location.
+    fn add_ptr(&mut self, to_add: usize) -> *mut u8;
+    /// Hash the head and tail of the bytes.
+    /// This is required to statisfy the `Hash` trait.
+    fn hash_head_tail<H: hash::Hasher>(
+        &self,
+        _byte_count: usize,
+        _state: &mut H,
+        _max_bytes_to_hash: usize,
+    ) {
+    }
+    /// Adjust the bytes to the specified alignment -- by default, this is a no-op.
+    fn adjust_to_align(self, _align: Align) -> Self {
+        self
+    }
+    /// Create an `AllocBytes` from a slice of `u8`.
+    fn from_bytes<'a>(slice: impl Into<Cow<'a, [u8]>>, _align: Align) -> Self;
+    /// Create an uninitialized `AllocBytes` of the specified size and alignment;
+    /// call the callback error handler if there is an error in allocating the memory.
+    fn uninit<'tcx, F: Fn() -> InterpError<'tcx>>(
+        size: Size,
+        _align: Align,
+        handle_alloc_fail: F,
+    ) -> Result<Self, InterpError<'tcx>>;
+}
+
+// Default `bytes` for `Allocation` is a `Box<[u8]>`.
+impl AllocBytes for Box<[u8]> {
+    fn uninit<'tcx, F: Fn() -> InterpError<'tcx>>(
```

Comment:
It's not uninit though, it returns a zeroed buffer?

Reply:
Huh that's true -- I just took this code from [here](https://github.com/rust-lang/rust/pull/100467/files#diff-1a9c0e976911962574444a7111463e6482e29f4dbb83697d513a7bdf155f67aeL236) where it was originally in the file

Reply:
The entire allocation is considered uninit based on its `InitMask`, yeah. But looking just at the bytes, which is the level of abstraction you have here, this is zeroed. The bytes cannot be uninit.